README.rst
A serializer for neomodel models.


.. image:: https://secure.travis-ci.org/efagerberg/neomodel-serializer.png
    :target: https://secure.travis-ci.org/efagerberg/neomodel-serializer/


.. image:: https://readthedocs.org/projects/neomodel-serializer/badge/?version=latest
    :alt: Documentation Status
    :scale: 100%
    :target: https://neomodel-serializer.readthedocs.io/en/latest/?badge=latest

Requirements
============

- Python 2.7, 3.4+
- neo4j 3.0, 3.1, 3.2, 3.3

CHANGELOG
Version 0.0.1.dev2
 * Adding the ability to serialize many entries, make sure there is a subfolder

Version 0.0.1.dev1
 * Starting version, includes base ability to serialize StructuredNode and StructuredRel

tests/conftest.py
from __future__ import print_function
import warnings
import os

from neomodel import config, db, clear_neo4j_database, change_neo4j_password
from neo4j.v1 import CypherError

warnings.simplefilter('default')

config.DATABASE_URL = os.environ.get(
    'NEO4J_BOLT_URL', 'bolt://neo4j:neo4j@localhost:7687')
config.AUTO_INSTALL_LABELS = True

try:
    clear_neo4j_database(db)
except CypherError as ce:
    # handle instance without password being changed
    if 'The credentials you provided were valid, but must be changed before you can use this instance' in str(ce):
        change_neo4j_password(db, 'test')
        db.set_connection('bolt://neo4j:test@localhost:7687')

        print("New database with no password set, setting password to 'test'")
        print("Please 'export NEO4J_BOLT_URL=bolt://neo4j:test@localhost:7687' for subsequent test runs")
    else:
        raise ce


tests/test_serializers.py
import json

from mock import patch

from neomodel import (
    db, clear_neo4j_database,
    ArrayProperty, IntegerProperty,
    StructuredNode, StructuredRel,
    StringProperty, RelationshipFrom
)
import pytest
from neomodel_serializer.serializers import StructuredThingSerializer


class Person(StructuredNode):
    name = StringProperty(required=True)


class ActorRel(StructuredRel):
    roles = ArrayProperty(StringProperty())


class Movie(StructuredNode):
    actors = RelationshipFrom(Person, 'ACTED_IN', model=ActorRel)
    directors = RelationshipFrom(Person, 'DIRECTED')

    released = IntegerProperty(required=True)
    tagline = StringProperty()
    title = StringProperty(unique_index=True, required=True)


@pytest.fixture()
def movie():
    yield Movie(title="Space Cop", released=2016).save()
    clear_neo4j_database(db)


@pytest.fixture()
def movies():
    for i in range(2):
        Movie(title=i, released=2016).save()
    clear_neo4j_database(db)


@pytest.fixture()
def mock_isinstance():
    path = 'neomodel_serializer.serializers.isinstance'
    with patch(path) as m:
        yield m


def test_parses_data(movie):
    """StructuredThingSerializer should parse data from instance."""
    p = Person(name="Mike Stoklasa").save()
    movie.directors.connect(p)
    actor_rel = movie.actors.connect(p, {'roles': ['Detective Ted Cooper']})

    expected = {
        'id': movie.id,
        'actors': [
            {'id': p.id,
             'name': p.name,
             'roles': actor_rel.roles}
        ],
        'directors': [
            {'id': p.id,
             'name': p.name}
        ],
        'released': movie.released,
        'tagline': movie.tagline,
        'title': movie.title,
    }
    actual = StructuredThingSerializer(movie).data
    assert actual == expected


def test_parses_only_certain_fields(movie):
    """"StructuredThingSerializer should exclude fields not
    declared in fields"""
    expected = {
        'title': movie.title,
        'released': movie.released,
    }
    actual = StructuredThingSerializer(
        movie, fields=('title', 'released')).data
    assert actual == expected


def test_serializes_structured_node_without_prop_or_rel_defs():
    """StructuredThingSerializer should handle a StructuredNode
    without any properties or relationship definitions."""
    class Foo(StructuredNode):
        pass

    x = Foo().save()
    expected = json.dumps({'id': x.id})
    actual = StructuredThingSerializer(x).serialized_data
    assert actual == expected


def test_raises_on_non_property_non_rel_def(movie, mock_isinstance):
    """StructuredThingSerializer should raise ValueError
    when a prop that is passed is not a Property or
    RelationshipDefinition"""
    mock_isinstance.return_value = False
    with pytest.raises(ValueError):
        StructuredThingSerializer(movie)


def test_serialize_many(movies):
    """StructuredThingSerializer should be able to serialize
    many movies"""
    qs = Movie.nodes
    expected = [{
        'id': movie.id,
        'actors': [],
        'directors': [],
        'released': movie.released,
        'tagline': movie.tagline,
        'title': movie.title,
    } for movie in qs]
    actual = StructuredThingSerializer(qs, many=True).data
    assert actual == expected


neomodel_serializer/serializers.py
from collections import OrderedDict
import json

from six import iteritems

from neomodel import Property
from neomodel.relationship_manager import RelationshipDefinition


class StructuredThingSerializer(object):
    """Initially made to only handle StructuredNodes
    but found that it also worked for StructuredRels."""

    def __init__(self, *args, **kwargs):
        """
        Args:
            instance (StructuredNode/StructuredRel): The thing to serialize.
            fields (str[], optional): Fields to include.
                                      Excludes subobject fields.
        """
        self.instance = args[0]
        self.fields = kwargs.get('fields', '__all__')
        self.many = kwargs.get('many')

        if self.many:
            self.data = []
            for i in self.instance:
                entry_data = StructuredThingSerializer(i).data
                self.data.append(entry_data)
        else:
            self.data = OrderedDict()
            props = self.instance.defined_properties()

            if self.should_include('id'):
                self.data['id'] = self.instance.id

            self.process_properties(props)
        self.serialized_data = json.dumps(self.data)

    def should_include(self, key):
        return key in self.fields or self.fields == '__all__'

    def process_properties(self, props):
        for prop in iteritems(props):
            self.process_property(prop)

    def process_property(self, prop):
        _, value = prop
        if isinstance(value, Property):
            self.process_neomodel_property(prop)
        elif isinstance(value, RelationshipDefinition):
            self.process_relationship_definition(prop)
        else:
            # If we get here, chances are neomodel has changed
            msg = "Unexpected property received: {}".format(prop)
            raise ValueError(msg)

    def process_neomodel_property(self, prop):
        key, _ = prop
        if self.should_include(key):
            self.data[key] = getattr(self.instance, key)

    def process_relationship_definition(self, prop):
        key, value = prop
        if self.should_include(key):
            rel_def = getattr(self.instance, key)
            self.process_rel_def_nodes(rel_def, key)

    def process_rel_def_nodes(self, rel_def, key):
        self.data[key] = []
        for node in rel_def:
            self.process_rel_def_node(rel_def, node, key)

    def process_rel_def_node(self, rel_def, node, key):
        rel_def_data = StructuredThingSerializer(node).data
        rels = rel_def.all_relationships(node)

        self.process_relationships(rels, rel_def_data, key)

    def process_relationships(self, rels, rel_def_data, key):
        for rel in rels:
            self.process_relationship(rel, rel_def_data, key)

        self.data[key].append(rel_def_data)

    def process_relationship(self, rel, rel_def_data, key):
        rel_data = StructuredThingSerializer(rel).data
        rel_data.pop('id')

        if rel_data:
            rel_def_data.update(rel_data)


test-with-docker-compose.sh
#!/usr/bin/env bash

# This script allows to run the test suite against all supported Python
# interpreters and neo4j versions on a host that has Docker Compose and
# Docker client installed.
# The script aborts on the first failure of any combination.
# It is not intended


write_compose_file () {
    cat > docker-compose.yml <<EOL
version: "3"

services:
  tests:
    image: "python:${PYTHON_VERSION}-alpine"
    volumes:
      - .:/src
      - pip-cache:/root/.cache/pip
    working_dir: /src
    command: sh -c "while ! nc neo4j 7687; do  sleep 1; done; python -B setup.py test"
    links:
      - neo4j
    environment:
      NEO4J_BOLT_URL: bolt://neo4j:neo4j@neo4j:7687
  neo4j:
    image: "neo4j:${NEO4J_VERSION}"
    environment:
      NEO4J_AUTH: none

volumes:
  pip-cache:
    driver: local
EOL
}


clean () {
    docker-compose rm --stop --force -v neo4j
    rm -f docker-compose.myl
}


rm -f ${dir}/**/*.pyc
    find ${dir} -name __pycache__ -exec rm -Rf {} \;

for NEO4J_VERSION in 3.0 3.1 3.2 3.3; do
    for PYTHON_VERSION in 2.7 3.4 3.5 3.6; do
        write_compose_file
        docker-compose up -d neo4j
        docker-compose up tests
        RESULT=$?
        if [ ${RESULT} != "0" ]; then
            echo "Tests with Python ${PYTHON_VERSION} against neo4j ${NEO4J_VERSION} failed. Stopping."
            clean
            exit ${RESULT}
        fi
    done
    clean
done

exit 0